pgrx/src/pg_catalog.rs

```diff
@@ -0,0 +1,1212 @@
+//LICENSE Portions Copyright 2019-2021 ZomboDB, LLC.
+//LICENSE
+//LICENSE Portions Copyright 2021-2023 Technology Concepts & Design, Inc.
+//LICENSE
+//LICENSE Portions Copyright 2023-2023 PgCentral Foundation, Inc. <contact@pgcentral.org>
+//LICENSE
+//LICENSE All rights reserved.
+//LICENSE
+//LICENSE Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+
+use crate as pgrx;
+use pgrx::datum::Array;
+use pgrx::pg_sys::Oid;
+use pgrx::pg_sys::Oid as Regproc;
+use std::ffi::{c_char, CStr};
+
+unsafe trait GetStruct<T> {
+    unsafe fn get_struct(raw: *const T) -> Self;
+}
+
+unsafe impl<T> GetStruct<T> for T {
+    unsafe fn get_struct(raw: *const T) -> Self {
+        unsafe { raw.read() }
+    }
+}
+
+unsafe impl GetStruct<pgrx::pg_sys::nameData> for &CStr {
+    unsafe fn get_struct(raw: *const pgrx::pg_sys::nameData) -> Self {
+        unsafe { CStr::from_ptr(raw.cast::<c_char>()) }
+    }
+}
+
+unsafe impl GetStruct<pgrx::pg_sys::int2vector> for &[i16] {
+    unsafe fn get_struct(raw: *const pgrx::pg_sys::int2vector) -> Self {
+        unsafe { (*raw).values.as_slice((*raw).dim1 as usize) }
+    }
+}
```

Comment:
I'm guessing the reason we need this is because sometimes we want to read the data, sometimes we want to get a CStr, and sometimes we need a slice, huh? But what about the other types that terminate in VLAs?

Reply:
All members in `FromData_*` are fixed-sized. There are only 3 exceptions: `oidvector`, `int2vector` and `bytea`. If there is `oidvector`, `int2vector` or `bytea` (only in `pg_largeobject`) in `FromData_*`, they must be the last struct member. Someone who needs `bytea` or there are more variable-sized types in future PostgreSQL versions could add it by themselves.

`nameData` is fixed-sized 64-byte array in fact. It's guaranteed that it's nul-padded so we can get `CStr` from it and we expose `&CStr` to users since it's more friendly.